chain/chain/src/validate.rs

```diff
@@ -171,8 +187,9 @@ pub fn validate_chunk_with_chunk_extra_and_receipts_root(
     }
 
     let gas_limit = prev_chunk_extra.gas_limit();
-    if chunk_header.gas_limit() < gas_limit - gas_limit / GAS_LIMIT_ADJUSTMENT_FACTOR
-        || chunk_header.gas_limit() > gas_limit + gas_limit / GAS_LIMIT_ADJUSTMENT_FACTOR
+    let adjustment_factor = gas_limit_adjustment_factor(gas_limit_adjustment_config);
+    if chunk_header.gas_limit() < gas_limit - gas_limit / adjustment_factor
+        || chunk_header.gas_limit() > gas_limit + gas_limit / adjustment_factor
```

Comment:
This is not the right place to perform this check. The chunk header should contain info from the prev chunk extra copied byte by byte. The adjustment should be deterministically computed as part of chunk application (be it in runtime or chain). An unchecked advise from me would be for you to grep for usages of ChunkExtra::new and/or ApplyChunkResult. You'll need to make the new gas limit part of the Result and propagate it into the Extra. 

* In the PR title you say that you want the adjustment to be dynamic. This means that you want to allow the chain to automatically adjust the gas limit based on some live data like gas usage.
* In the PR summary you say that you need it for testing and benchmarking. In this case it seems like it should be sufficient to just have a static config for this value. This would greatly simplify the logic and you wouldn't need to go into chain, validation, etc. 

Which of the above are you aiming for?

Reply:
> This is not the right place to perform this check.

Can this be considered a preexisting issue? Since that's where the check currently is. This PR doesn't add or modify the check, it "just" adds configuration which allows overriding the hardcoded `GAS_LIMIT_ADJUSTMENT_FACTOR`.

> Which of the above are you aiming for?

My understanding is that it's both. For now the main use-case is starting a node for a continuous benchmark and let it automatically reach the highest gas limit that it can handle. Since one of the aims of these benchmarks is to uncover what could be possible in terms of transactions per second. More details are given in #11460.

> for testing and benchmarking. In this case it seems like it should be sufficient to just have a static config for this value.

Yes, this can be done by modifying `gas_limit` in `genesis.json` or setting another [INITIAL_GAS_LIMIT](https://github.com/near/nearcore/blob/f147992c373b333e6435a31bab7eb32fbb8f1f30/core/chain-configs/src/lib.rs#L49). Though it's tedious to determine the right value. Also this value depends on the hardware that is used and how `nearcore` performance characteristics change. Hence the *dynamic* gas limit adjustments which aim to make the node determine the "ideal" gas limit automatically.

> An unchecked advise from me would be for you to grep for usages of ChunkExtra::new and/or ApplyChunkResult. You'll need to make the new gas limit part of the Result and propagate it into the Extra.

Thanks! I'll look into it as I'm trying to find out if there is another way to modify the gas limit which will not make this test fail:

https://github.com/near/nearcore/blob/02fa4d2ec6c5d30702a403f25803e5af2708f993/chain/chain/src/validate.rs#L157-L159

Reply:
@wacban this looks like the right place to me. Unlike congestion control, this adjustment is meant to be subjective for every node, that's even its original intent. So we reach both goals here.

The reason is - if gas limit becomes not enough, nodes can nudge it to the more convenient value. If nodes do that, it's impossible to record in chunk extra for prev chunk, final gas limit can be known only when chunk is received. Exact gas limit changes were never coded before, though. Thus #11958 is correct.

Reply:
Thanks Alex! 

Yeah I stand corrected it may be ok to do it here. I suppose the idea is to leave it to the chunk producer to adjust the limit as they see fit. Given this is only meant for testing it seems fine. 

My only concern with regard to this implementation is that it will introduce divergence between the info in ChunkExtra and the ChunkHeader of the next block. I would suggest going through the usages of the gas limit in the codebase and make sure it's always obtained from the header and not from the extra. Alternatively you can consider updating the extra after the header is accepted. 

TBH I'm still on the edge about making this purely testing change a part of the protocol - even if disabled.

Reply:
@Longarithm, @wacban thanks for the suggestions! I can look into it when I'm back next week.

> TBH I'm still on the edge about making this purely testing change a part of the protocol - even if disabled.

What are the next steps in making that decision? As I should avoid working on something that would be obsolete.

Reply:
@mooori Can you check if there are any sensible out of protocol alternatives? Perhaps the same goal can be achieved by the testing framework or by controlling the traffic?